libp2p/transports/tls/certificate.nim

```diff
@@ -0,0 +1,489 @@
+# Nim-LibP2P
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import std/[sequtils, strutils]
+
+import stew/byteutils
+import chronicles
+
+import mbedtls/pk
+import mbedtls/ctr_drbg
+import mbedtls/entropy
+import mbedtls/ecp
+import mbedtls/sha256
+import mbedtls/md
+import mbedtls/asn1
+import mbedtls/asn1write
+import mbedtls/x509
+import mbedtls/x509_crt
+import mbedtls/oid
+import mbedtls/debug
+import mbedtls/error
+import nimcrypto/utils
+import ../../crypto/crypto
+
+logScope:
+  topics = "libp2p tls certificate"
+
+# Constants and OIDs
+const
+  P2P_SIGNING_PREFIX = "libp2p-tls-handshake:"
+  SIGNATURE_ALG = MBEDTLS_MD_SHA256
+  EC_GROUP_ID = MBEDTLS_ECP_DP_SECP256R1
+  LIBP2P_EXT_OID_DER: array[10, byte] =
+    [0x2B, 0x06, 0x01, 0x04, 0x01, 0x83, 0xA2, 0x5A, 0x01, 0x01]
+    # "1.3.6.1.4.1.53594.1.1"
+
+# Exception types for TLS certificate errors
+type
+  TLSCertificateError* = object of Exception
+  ASN1EncodingError* = object of TLSCertificateError
+  KeyGenerationError* = object of TLSCertificateError
+  CertificateCreationError* = object of TLSCertificateError
+  CertificateParsingError* = object of TLSCertificateError
+  IdentityPubKeySerializationError* = object of TLSCertificateError
+  IdentitySigningError* = object of TLSCertificateError
+
+# Define the P2pExtension and P2pCertificate types
+type
+  P2pExtension* = object
+    publicKey*: seq[byte]
+    signature*: seq[byte]
+
+  P2pCertificate* = object
+    certificate*: mbedtls_x509_crt
+    extension*: P2pExtension
+
+proc ptrInc*(p: ptr byte, n: uint): ptr byte =
+  ## Utility function to increment a pointer by n bytes.
+  cast[ptr byte](cast[uint](p) + n)
+
+proc generateSignedKey(
+    signature: seq[byte], pubKey: seq[byte]
+): seq[byte] {.raises: [ASN1EncodingError].} =
+  ## Generates the ASN.1-encoded SignedKey structure.
+  ##
+  ## The SignedKey structure contains the public key and its signature,
+  ## encoded as a SEQUENCE of two OCTET STRINGs.
+  ##
+  ## Parameters:
+  ## - `signature`: The signature bytes.
+  ## - `pubKey`: The public key bytes.
+  ##
+  ## Returns:
+  ## A sequence of bytes representing the ASN.1-encoded SignedKey.
+  ##
+  ## Raises:
+  ## - `ASN1EncodingError` if ASN.1 encoding fails.
+  const extValueSize = 256 # Buffer size for ASN.1 encoding
+  var
+    extValue: array[extValueSize, byte]
+    extPtr: ptr byte = addr extValue[extValueSize - 1] # Start at the end of the buffer
```

Comment:
At first I thought it was an error, I had to check `mbedtls_asn1_write_octet_string` to understand that this function works backward. Maybe you can add this as a comment somewhere.

Reply:
done.